src/Raven.Server/Documents/Queries/StreamJsonlDocumentQueryResultWriter.cs

```diff
@@ -0,0 +1,67 @@
+﻿using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Server.Json;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Queries
+{
+    public class StreamJsonlDocumentQueryResultWriter : IStreamQueryResultWriter<Document>
+    {
+        private readonly AsyncBlittableJsonTextWriter _writer;
+        private readonly JsonOperationContext _context;
+
+        public StreamJsonlDocumentQueryResultWriter(Stream stream, JsonOperationContext context)
+        {
+            _context = context;
+            _writer = new AsyncBlittableJsonTextWriter(context, stream);
+        }
+
+        public ValueTask DisposeAsync()
+        {
+            return _writer.DisposeAsync();
+        }
+
+        public void StartResponse()
+        {
+        }
+
+        public void StartResults()
+        {
+        }
+
+        public void EndResults()
+        {
+        }
+
+        public async ValueTask AddResultAsync(Document res, CancellationToken token)
+        {
+            _writer.WriteDocument(_context, res, metadataOnly: false);
+            _writer.WriteNewLine();
+            await _writer.MaybeFlushAsync(token);
+        }
+
+        public void EndResponse()
+        {
+        }
+
+        public ValueTask WriteErrorAsync(Exception e)
+        {
+            throw new NotSupportedException();
+        }
+
+        public ValueTask WriteErrorAsync(string error)
+        {
+            throw new NotSupportedException();
+        }
+
+        public void WriteQueryStatistics(long resultEtag, bool isStale, string indexName, long totalResults, DateTime timestamp)
+        {
+            throw new NotSupportedException();
```

Comment:
For streaming, we sent this at the _start_ of the operation. That can be done via the first item (that the client will know to ignore? Not sure that I like that, to be honest. Maybe we can send that as headers?

Reply:
I was thinking about sending stats at the start of the operation, also was not sure what to make of it. I think the headers approach might be better now that you mention it. I guess the choice is down to what is easier for the clients to implement?

To add a second use-case as a potential tie-breaker, i was hoping to be able to pipe results directly into an another stream - say we have filtering/projections on raven and we want to respond directly back to a client over a RESTful API.

Another thought that comes to mind is that if we have to introduce a wrapper around individual items, then introducing headers as a second way to provide data might not be a good choice.